File: solutions/go/lasagna-master/2/lasagna_master.go
package lasagna

func PreparationTime(layers []string, averageTime int) int {
    if averageTime == 0 {
        averageTime = 2
    }
    return len(layers) * averageTime
}

func Quantities(layers []string) (noodles int, sauce float64) {
    for _, layer := range(layers) {
        switch layer {
        case "noodles":
            noodles += 50
        case "sauce":
            sauce += 0.2
        }
    }
    return
}

func AddSecretIngredient(yours []string, mine []string) {
    if len(mine) > 1 && len(yours) > 1 {
        mine[len(mine) - 1] = yours[len(yours) - 1]
    }
}

func ScaleRecipe(quantities []float64, portions int) (ret []float64) {
    ret = make([]float64, len(quantities))
    for i := range(ret) {
        ret[i] = quantities[i] * float64(portions) / 2
    }
    return 
}

File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

#include <algorithm>

namespace lasagna_master {

int preparationTime(const std::vector<std::string>& layers, int n) {
    return n * layers.size();
}

amount quantities(const std::vector<std::string>& layers) {
    amount a{};
    a.noodles = std::count(layers.begin(), layers.end(), "noodles") * 50;
    a.sauce = std::count(layers.begin(), layers.end(), "sauce") * 0.2;
    return a;
}
    
void addSecretIngredient(std::vector<std::string>& my_recipe, const std::vector<std::string>& friends_recipe) {
    my_recipe.back() = friends_recipe.back();
}
    
std::vector<double> scaleRecipe(const std::vector<double>& two_portions, int n) {
    std::vector<double> result(two_portions.size());
    std::transform(two_portions.begin(), two_portions.end(), result.begin(), [&n](const double& for_two){ return for_two / 2 * n; });
    return result;
}
    
void addSecretIngredient(std::vector<std::string>& my_recipe, const std::string secret) {
    my_recipe.back() = secret;    
}

}  // namespace lasagna_master
